### PR TITLE
also load extension from environment

### PR DESCRIPTION
**Pull request by Sasasu**
review request @higuoxing @beeender

### CODE REVIEW

File: gpMgmt/bin/lib/gpcreateseg.sh [hunk 1]
```diff
@@ -93,7 +93,11 @@ CREATE_QES_PRIMARY () {
     LOG_MSG "[INFO][$INST_COUNT]:-Start Function $FUNCNAME"
     LOG_MSG "[INFO][$INST_COUNT]:-Processing segment $GP_HOSTADDRESS"
     # build initdb command
-    cmd="$EXPORT_LIB_PATH;$INITDB"
+    cmd=""
+    if [ x"$PRELOAD_LIBRARIES" != x"" ]; then
+        cmd="$cmd GP_PRELOAD_LIBRARIES=$PRELOAD_LIBRARIES;"
+    fi
+    cmd="$EXPORT_LIB_PATH;$INITDB" # QQQ TODO
```
Comment (beeender) [inline gpMgmt/bin/lib/gpcreateseg.sh]: What to do?

File: gpMgmt/bin/gpinitsystem [hunk 2]
```diff
@@ -1182,6 +1182,10 @@ CREATE_QD_DB () {
 
 
 		# build initdb command
+		cmd=""
+		if [ x"$PRELOAD_LIBRARIES" != x"" ]; then
```
Comment (beeender) [inline gpMgmt/bin/gpinitsystem:1186]: Why do we need to check `PRELOAD_LIBRARIES`? Who defines this? I am bit confused.

File: gpMgmt/bin/lib/gpcreateseg.sh [hunk 3]
```diff
@@ -93,7 +93,11 @@ CREATE_QES_PRIMARY () {
     LOG_MSG "[INFO][$INST_COUNT]:-Start Function $FUNCNAME"
     LOG_MSG "[INFO][$INST_COUNT]:-Processing segment $GP_HOSTADDRESS"
     # build initdb command
-    cmd="$EXPORT_LIB_PATH;$INITDB"
+    cmd=""
+    if [ x"$PRELOAD_LIBRARIES" != x"" ]; then
```
Comment (beeender) [inline gpMgmt/bin/lib/gpcreateseg.sh:97]: ditto

File: src/backend/utils/init/miscinit.c [hunk 4]
```diff
@@ -1688,6 +1688,17 @@ process_shared_preload_libraries(void)
...
 
+	if ((shared_preload_libraries_string == NULL || strlen(shared_preload_libraries_string) == 0)
+		&& (IsBootstrapProcessingMode() || IsInitProcessingMode()))
+	{
+		/*
+		 * if we are in init or bootstrap mode. postgresql.conf does not exists yet.
+		 * try to load libraries from os environment
+		 */
+		char * const env_preload = getenv("GP_PRELOAD_LIBRARIES");
+		load_libraries(env_preload, "GP_PRELOAD_LIBRARIES", false);
```
Comment (beeender) [inline src/backend/utils/init/miscinit.c:1699]: Maybe we can rename it to `GP_INIT_PRELOAD_LIBRARIES` to make it clear it only works during the init process.

File: src/backend/utils/init/miscinit.c [hunk 5]
```diff
@@ -1688,6 +1688,17 @@ process_shared_preload_libraries(void)
 {
 	process_shared_preload_libraries_in_progress = true;
 
+	if ((shared_preload_libraries_string == NULL || strlen(shared_preload_libraries_string) == 0)
+		&& (IsBootstrapProcessingMode() || IsInitProcessingMode()))
```
Comment (beeender) [inline src/backend/utils/init/miscinit.c:1692]: just curious, for bootstrap and init, the `shared_preload_libraries_string` should always be null or empty.
probably an assertion is better. No strong opinions on this.

File: gpMgmt/bin/lib/gpcreateseg.sh (hunk 1 above)
Comment (Sasasu) [reply gpMgmt/bin/lib/gpcreateseg.sh]: A mark to mark this part needs to be modified. 
only used by me, removed

**Review by higuoxing — COMMENTED** [review]
Instead of passing the `preload-libraries` to the postmaster by an environment variable `GP_PRELOAD_LIBRARIES`, can we add a new commandline option `--preload-libraries` to `postgremaster` and pass it via `initdb`?

If we specify the `GP_PRELOAD_LIBRARIES` environment variable and didn't specify the `preload-libraries` argument of the `gpinitsystem`, the coordinator's postmaster will load the library while QE's postmasters will not.

**Comment by Sasasu** [comment]
move to https://github.com/Sasasu/gpdb/pull/2

**Pull request closed without merging**